[lldb/Target] Add SyntheticFrameProvider class

medismailben opened the pull request:
This patch introduces a new way to reconstruct the thread stackframe
list.

New `SyntheticFrameProvider` classes can lazy fetch a StackFrame at
index using a provided StackFrameList.

In can either be the real unwinder StackFrameList or we could also chain
SyntheticFrameProviders to each others.

This is the foundation work to implement ScriptedFrameProviders, which
will come in a follow-up patch.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+
```
Spurious newline

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+
+//===------------------------------------------------------------*- C++ -*-===//
```
```suggestion
//===----------------------------------------------------------------------===//
```

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+
+namespace lldb_private {
+
+/// Descriptor for configuring a synthetic frame provider.
```
This is kind of redundant.

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+/// Descriptor for configuring a synthetic frame provider.
+///
+/// This struct contains the metadata needed to instantiate a frame provider
+/// and optional filters to control which threads it applies to.
```
This is also pretty vague and seems very "Scripted*" specific. If I'm implementing a C++ based SyntheticFrameProviderDescriptor (e.g. for Python frames), wy do I need to provided scripted metadata? From a design POV, I would expect the constructor to take the stuff that's encapsulated here, and this is part of the "Scripted*" bridging.

medismailben replied:
Ok, I made a different callback for "native" frame providers that doesn't require SyntheticFrameProviderDescriptor. However I still need this in lldb core since the targets needs to hold them, in case it doesn't have a process (and threads) yet. I left it here but I could move it to Utility next to ScriptedMetadata if you prefer.

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+  ///     otherwise an \a llvm::Error.
+  static llvm::Expected<lldb::SyntheticFrameProviderSP>
+  CreateInstance(lldb::StackFrameListSP input_frames,
+                 const SyntheticFrameProviderDescriptor &descriptor);
```
See my point above, I think this class should take a `thread_specs` instead, and you can do the bridging with the descriptor in the ScriptedFrameProvider

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,133 @@
+protected:
+  SyntheticFrameProvider(lldb::StackFrameListSP input_frames);
+
+  lldb::StackFrameListSP m_input_frames;
```
[bikeshedding]`m_original_frames`?[/bikeshedding]

JDevlieghere commented on lldb/include/lldb/Target/SyntheticFrameProvider.h:
```diff
@@ -0,0 +1,163 @@
+    if (thread_specs.empty())
+      return true;
+
+    // Check if the thread matches any of the specs (OR logic)
```
```suggestion
    // Check if the thread matches any of the specs (OR logic).
```

medismailben committed:
[lldb/Target] Add SyntheticFrameProvider class

This patch introduces a new way to reconstruct the thread stackframe
list.

New SyntheticFrameProvider classes can lazy fetch a StackFrame at
index using a provided StackFrameList.

In can either be the real unwinder StackFrameList or we could also chain
SyntheticFrameProviders to each others.

This is the foundation work to implement ScriptedFrameProviders, which
will come in a follow-up patch.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

JDevlieghere reviewed (approved):
LGTM!
